src/contexts/auth/AuthContext.tsx

```diff
@@ -52,97 +53,102 @@ export const useAuthActions = (): AuthActions => {
+  const [isLoggedInLocally, setIsLoggedInLocally] = usePersistedState(
+    USER_LOGGED_IN_LOCAL_STORAGE_KEY,
+    false
   );
 
+  const { pushToast } = useToastActions();
+
   /**
-   * Only sets the user state to logged out.
+   * Sets the user state to logged out and clears local storage
    */
-  const setLoggedOut = useCallback(() => {
-    setAuthState(LOGGED_OUT);
-  }, []);
+  const setLoggedOut = useCallback(
+    (isUnauthorized = false) => {
+      if (isUnauthorized) {
+        pushToast(EXPIRED_SESSION_MESSAGE);
+      }
+
+      setAuthState(LOGGED_OUT);
+      setUserSigninAddress('');
+      /**
+       * NOTE: clearing localStorage completely will also clear the SWR cache, which
+       * will require users to re-fetch data if they visit a profile they've already
+       * visited (including their own). In the future, we should clear data more
+       * selectively (such as only sensitive data)
+       */
+      clearLocalStorageWithException([]);
+    },
+    [pushToast, setUserSigninAddress]
+  );
 
   /**
-   * Fully logs user out and clears storage.
+   * Fully logs user out by calling the logout endpoint and logging out in app state
    */
-  const logOut = useCallback(() => {
+  const logOut = useCallback(async () => {
+    await _fetch('/auth/logout', 'logout', { body: {} });
     setLoggedOut();
-    setToken('');
-    setUserId('');
-    setUserSigninAddress('');
-    /**
-     * NOTE: clearing localStorage completely will also clear the SWR cache, which
-     * will require users to re-fetch data if they visit a profile they've already
-     * visited (including their own). In the future, we should clear data more
-     * selectively (such as only sensitive data)
-     */
-    clearLocalStorageWithException([]);
-  }, [setLoggedOut, setToken, setUserId, setUserSigninAddress]);
-
-  const logIn = useCallback(
-    async (payload: LoggedInState, address: string) => {
+  }, [setLoggedOut]);
+
+  const setLoggedIn = useCallback(
+    async (address: string) => {
       try {
-        setToken(payload.jwt);
-        setUserId(payload.userId);
-        setAuthState(payload);
+        setAuthState(LOGGED_IN);
         setUserSigninAddress(address.toLowerCase());
       } catch (error: unknown) {
         const errorMessage = error instanceof Error ? error.message : 'Unknown error';
-        logOut();
+        setLoggedOut();
         throw new Error('Authorization failed! ' + errorMessage);
       }
     },
-    [logOut, setToken, setUserId, setUserSigninAddress]
+    [setLoggedOut, setUserSigninAddress]
   );
 
   const setStateToLoading = useCallback(() => {
     setAuthState(LOADING);
   }, []);
 
-  const { pushToast } = useToastActions();
+  const handleUnauthorized = useCallback(() => {
+    setLoggedOut(true);
+  }, [setLoggedOut]);
+
+  useEffect(() => {
+    setIsLoggedInLocally(authState === LOGGED_IN);
+  }, [authState, setIsLoggedInLocally]);
```

Comment:
a bit concerned there will be a race condition between this effect and the one below.

according to the line above, `authState` will always begin as `UNKNOWN`:
```js
const [authState, setAuthState] = useState<AuthState>(UNKNOWN);
```

this means this effect could at first call `setIsLoggedInLocally(UNKNOWN === LOGGED_IN)`, which results in `false`, even though the other useEffect may end up determining that the `authState` is `LOGGED_IN`.

the worst case is that the user's previous local state is logged in, but this effect wipes it out and sets it to false, and we're unable to determine whether they were logged in before.

could we just remove this effect and call `setIsLoggedInLocally` in the below effect? (commented there)

Reply:
makes sense! yeah totally valid concern, good catch.
I originally moved it out into a separate useEffect because I was seeing the metamask sign message open twice and thought that something in here was causing it to run more than once. I didn't see it anymore after I moved it so I thought that did the trick

But I just tried moving it all back into the same one  and it looks like it's behaving as expected still  💪